In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

#checks for missing values 
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

This missing value dataset shows that id has 1 missing value and song has 11 missing values.

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

#this df.isna().sum() checks for missing values
df.isna().sum()
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

The DataFrame has 391 rows and 11 columns.


General definitions

Observations: In a dataset a observation refers to a single data point or record in a dataset- Each observation is typically represented by a row in a table or data frame. 

Variables: Charectersitics or properties that describe observations - Each variable is represented by a column in a table or dataframe.


Definitions in context of my dataset

Observations: In my dataset each row represents an observation - each observations corresponds to a villager from the game animal crossing - each row gives information about one villager.
Variables: each column in my dataset represents a variable that describes some aspect of the villagers for eg : name, species, personality, birthday song etc are all variables associated with describing some aspect of each villager. 

-> In the animal crossing dataset each row(observation) represents one villager and each column(variable) represents or describes a specific charecteristic of that villager.


In [4]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

#general summary statistics for numerical columns 
summary_stats = df.describe()
df.describe()

The DataFrame has 391 rows and 11 columns.


row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [7]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

species_counts = df['species'].value_counts()
print(species_counts)

The DataFrame has 391 rows and 11 columns.
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64


# 4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

In [8]:
import pandas as pd

#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

#Check for missing values 
df.isna().sum()


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [9]:
import pandas as pd

#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

#Check how many rows and columns are in the dataset 
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

The DataFrame has 891 rows and 15 columns.


In [10]:
import pandas as pd

#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Disperencies between size of the dataset given by df.shape() and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column. 

df.shape() shows the total number of of columns both numeric and non-numeric 
df.describe() only summarizes numeric columns not non-numeric - therefore shows fewer columns than df.shape() indicates 

(a) the number of columns it analyzes 
     - df.shape() tells me that there are 15 columns 
     - df.describe() only shows me 6 numeric columns : survived, pclass, age, sibsp, parch, and fare. 
     
(b) The total number of rows are 891 given by the df.shape() attribute however using the method df.describe() we figure out that for all columns from survived to fare the count is 891, however for age the count is 714 indicating missing age values. 891-714 gives us 177 indicating 177 missing values. 
    The difference between the two is that df.describe() only provides summary statistics for columns with numeric data therefore if a column has missing values its count will be less than the total number of rows reflecting those missing values. 
    


# 5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

an "attribute", such as df.shape which does not end with ()
and a "method", such as df.describe() which does end with ()

Attributes 
     attributes are values or properties associated with an object, they are accessed directly without parantheses they provide information about an objects or its state
     
Methods 
    methods are functions defined within a class that perform actions or operations on the object - they are called using parantheses, they often perform computations or actions and may return results or modify the state of the object. 
    Attributes: Provide information or properties of an object and are accessed without parentheses. Example: df.shape gives the dimensions of the DataFrame.
    Methods: Perform actions or operations on an object and are called with parentheses. Example: df.describe() computes and returns summary statistics.
    
In conclusion attributes give information about an object or its state property etc whereas methods perform actions or operations on the object the major difference while coding attributes and methods is that methods are called with parentheses for eg df.describe() whereas attributes are accessed directly without using parantheses.


# Summary of Chatbot Session

Here's a summary of our chat:

Pandas DataFrame Rows and Columns:

You asked how to determine the number of rows and columns in a pandas DataFrame. We used the df.shape attribute to get this information, which returns a tuple representing (number of rows, number of columns).
Definitions of "Observations" and "Variables":

We discussed the general definitions of observations (rows in the dataset, representing individual data points) and variables (columns, representing characteristics of the observations). In the context of your dataset (Animal Crossing villagers), each villager is an observation, and each column (like species, personality) is a variable.
Summarizing the Dataset:

You used df.describe() to summarize numerical columns and df['column'].value_counts() to summarize categorical columns (like species or personality). This provides both statistical insights for numerical data and frequency counts for categorical data.
We discussed how the simplest summarization involves using these two methods to quickly understand the distribution of values.
Difference Between print(summary_stats) and df.describe():

We clarified that df.describe() computes and immediately shows the summary statistics, while print(summary_stats) displays the result of df.describe() after saving it in a variable.
Difference Between Attributes and Methods:

You learned that attributes (like df.shape) provide information about an object without requiring parentheses, while methods (like df.describe()) perform actions and require parentheses.
The discussion mainly focused on understanding basic pandas functionality and how to interpret dataset summaries using common methods and attributes.

url to chatbot session: https://chatgpt.com/share/b822d6a6-4fa5-4a9e-8409-727e343a8404